### PR TITLE
Support empty variants

### PR DESCRIPTION
**Pull request by objmagic**
See [Mantis 7583](https://caml.inria.fr/mantis/view.php?id=7583) for discussions.

This PR allows one to declare type with zero variants:

```
type t = |
let f (x:t) = match x with _ -> .
```

I added a variant `Empty_pattern` in `error` type so that implementation can be easier and cleaner, although user will never encounter it (its error message).

### CODE REVIEW

**Review by Drup — APPROVED** [review]
> I added a variant Empty_pattern in error type so that implementation can be easier and cleaner, although user will never encounter it (its error message).

having looked at the code, I got very confused at why this was an error, so I believe it would be better to keep an `assert false` there.

The rest is very simple and quite straightforward, since absurd patterns are already there. 

This code is accepted, I do not know if this is good or bad.
```ocaml
type t = | 
let f : t -> int = function _ -> 3 ;;
```

**Comment by Drup** [comment]
Also, I think you should add a test that the following code passes without triggering exhaustiveness warning (currently it does, and it is cool :p).
```ocaml
type t = | 
let f : t option -> int = function None -> 3 
```

**Comment by Octachron** [comment]
@Drup :  I am not sure if the behavior in your first example is surprising, since the same behavior already happens for any empty type: either
```Ocaml
type (_,_) eq = Eq:('a,'a) eq
type empty = (int,float) eq
```
or
```OCaml
type absurd  = {x: 'a. 'a }
```
yields the following unusable but accepted functions

```OCaml
let f: empty -> int = function _ -> 3
let f : absurd -> int = function _ -> 3
```

**Comment by Drup** [comment]
@Octachron It isn't surprising. However, the typechecker knows that my `t` or your `empty` is not inhabited, and could use that information to warn the user. The typechecker doesn't know that `absurd` is not inhabited.

Let's call that a potential feature request for our contributors that enjoy making good errors and warnings. :D

**Comment by gasche** [comment]
The current design of the pattern-matching codebase is untyped (it does not take the type of the rows as input), which means that detecting that this case is unused is not an easy change. Having a design that propagates type information is reasonable in theory, and we might want to look at it eventually, but that's not a short-term change.

**Comment by Drup** [comment]
@gasche Well, the exhaustiveness check does take type information into account. doesn't it ? I suppose this is done separately.

**Comment by garrigue** [comment]
@gasche, @Drup The exhaustiveness check does a second pass on the missing cases, checking whether they are really inhabited using the type checker. I would expect it to ensure exhaustiveness here for free. If this is not the case, something has been overlooked.

**Comment by Drup** [comment]
@garrigue Exhaustiveness does work here, as you expect. It's reachability that is not complete. Given the presence of empty types, reachability of wildcards can benefits from type information just like exhaustiveness, but this is not done (yet!).

All of this is not necessary to make this feature useful, and since the only opposition on the Mantis Issue is regarding the difficulty of implementing empty records, I propose we merge!

**Comment by objmagic** [comment]
@Drup Thanks for your review. I have address two of your comments --- add a test and use `assert false`

File: typing/typecore.ml [hunk 1]
```diff
@@ -97,6 +97,7 @@ type error =
   | Illegal_letrec_expr
   | Illegal_class_expr
   | Unbound_value_missing_rec of Longident.t * Location.t
+  | Empty_pattern
```
Comment (Drup) [inline typing/typecore.ml:100]: We don't need the whole `Empty_pattern` thing anymore, right ?

File: typing/typecore.ml [hunk 2]
```diff
@@ -1044,7 +1045,11 @@ and type_pat_aux ~constrs ~labels ~no_existentials ~mode ~explode ~env
         pat_env = !env }
       in
       if explode > 0 then
-        let (sp, constrs, labels) = Parmatch.ppat_of_type !env expected_ty in
+        let (sp, constrs, labels) =
+          try
+            Parmatch.ppat_of_type !env expected_ty
+          with Parmatch.Empty -> raise (Error (loc, !env, Empty_pattern))
```
Comment (objmagic) [inline typing/typecore.ml:1051]: @Drup Seems I misunderstood your review. Exception `raise Error...` sometimes is used to backtrack in this function. So I have to add a new variant `Empty_pattern` in `error` to use `raise Err...`

Comment (Drup) [reply typing/typecore.ml:1051]: I see, that's a bit inconvenient, but I don't have a better proposition, so I suppose it's fine.

**Comment by objmagic** [comment]
@garrigue I would appreciate if you'd like to do a round of review of this PR.

**Review by garrigue — CHANGES_REQUESTED** [review]
The change seems fine, and has already been approved at the November developer's meeting.
Note however that this can break lots of stuff, so you need to test all the tools in the distribution, in particular ocamldoc, and also you need to update the manual to reflect this change.

File: manual/manual/refman/typedecl.etex [hunk 3]
```diff
@@ -31,6 +31,7 @@ type-equation:
 type-representation:
           '=' ['|'] constr-decl { '|' constr-decl }
         | '=' record-decl
+        | '|'
```
Comment (garrigue) [inline manual/manual/refman/typedecl.etex]: missing `=` ?

**Comment by objmagic** [comment]
Added the missing equal sign.

I tested `-dparsetree`, `-dtypedtree`, `-dsource`, `-i`, and ocamldoc (html and latex). They all work as expected. If I am missing something here, please feel free to point out.

**Comment by garrigue** [comment]
@Drup Actually it's not that it's not implemented for redundancy, but that it is explicitly disabled when there is only one case: `(not refute && pref = [])`. IIRC, the main goal was to avoid warnings for other syntactic constructs, such as let definitions or function formal arguments. We might either remove this restriction, or refine it according to the syntactic form (but that information is not propagated currently).

File: typing/parmatch.ml [hunk 4]
```diff
@@ -897,7 +897,7 @@ let pat_of_constr ex_pat cstr =
 let orify x y = make_pat (Tpat_or (x, y, None)) x.pat_type x.pat_env
 
 let rec orify_many = function
-| [] -> assert false
+| [] -> raise Empty
```
Comment (garrigue) [inline typing/parmatch.ml]: Looking back at this code, I'm concerned by this exception.
`orify_many` is called in many places, and are you sure that they are all ready to catch `Empty`?
My feeling is that you should rather modify `ppat_of_type` to raise an exception when `pats = []`, without calling `orify_many` on the empty list. And this should be documented in the .mli.

Comment (garrigue) [inline typing/parmatch.ml]: Looking back at this code, I'm concerned by this exception.
`orify_many` is called in many places, and are you sure that they are all ready to catch `Empty`?
My feeling is that you should rather modify `ppat_of_type` to raise an exception when `pats = []`, without calling `orify_many` on the empty list. And this should be documented in the .mli.

**Review by garrigue — CHANGES_REQUESTED** [review]
Sorry for the delayed request.

File: typing/parmatch.ml (hunk 4 above)
Comment (objmagic) [reply typing/parmatch.ml]: Good points. I realized this when I finished implementation. I checked all call sites and thought it should be fine.

Obviously it is better to raise in `ppat_of_type`. I have pushed the change.

Comment (objmagic) [reply typing/parmatch.ml]: I have also added document in .mli

**Review by garrigue — APPROVED** [review]

**Comment by garrigue** [comment]
Just merged the PR. Thank you for all your revisions.

**Comment by fpottier** [comment]
I note that this PR allows `type t = | | A of int`, which is taken to mean the same thing as `type t = | A of int`. In my eyes, this is harmless, but undesirable. I propose to fix it as part of the parser cleanup on which I am currently working.

**Comment by gasche** [comment]
Yep, fine with me!

**Pull request merged**